esphome/components/voice_kit/__init__.py

```diff
@@ -1,31 +1,113 @@
+import hashlib
+from pathlib import Path
 import esphome.codegen as cg
 import esphome.config_validation as cv
 from esphome import pins
+from esphome import automation, core, external_files
 from esphome.components import i2c
-from esphome.const import CONF_ID, CONF_RESET_PIN
+from esphome.const import (
+    CONF_ID,
+    CONF_RAW_DATA_ID,
+    CONF_RESET_PIN,
+    CONF_URL,
+    CONF_VERSION,
+)
+from esphome.core import HexInt
 
 CODEOWNERS = ["@kbx81"]
 DEPENDENCIES = ["i2c"]
 
+CONF_FIRMWARE = "firmware"
+
+DOMAIN = "voice_kit"
+
 voice_kit_ns = cg.esphome_ns.namespace("voice_kit")
 VoiceKit = voice_kit_ns.class_("VoiceKit", cg.Component, i2c.I2CDevice)
+VoiceKitFlashAction = voice_kit_ns.class_("VoiceKitFlashAction", automation.Action)
+
+
+def _compute_local_file_path(url: str) -> Path:
+    h = hashlib.new("sha256")
+    h.update(url.encode())
+    key = h.hexdigest()[:8]
+    base_dir = external_files.compute_local_file_dir(DOMAIN)
+    return base_dir / key
+
+
+def download_firmware(config):
+    url = config[CONF_URL]
+    path = _compute_local_file_path(url)
+    external_files.download_content(url, path)
+
+    return config
+
 
 CONFIG_SCHEMA = (
     cv.Schema(
         {
             cv.GenerateID(): cv.declare_id(VoiceKit),
             cv.Required(CONF_RESET_PIN): pins.gpio_output_pin_schema,
+            cv.GenerateID(CONF_RAW_DATA_ID): cv.declare_id(cg.uint8),
+            cv.Optional(CONF_FIRMWARE): cv.All(
+                {
+                    cv.Required(CONF_URL): cv.url,
+                    cv.Required(CONF_VERSION): cv.version_number,
```

Comment:
I strongly think that we also need to force an md5 (or similar hash) to verify the binary file we are about to embed is fully correct.

Reply:
Done in [`392ea65` (#68)](https://github.com/esphome/voice-kit/pull/68/commits/392ea651dff87d1d335844634cb34f12ac0d13f5)